qemu/machine.py

```diff
@@ -418,31 +424,66 @@ def fill_source(self):
                     )
                 else:
                     self.use_type_name("DeviceState")
-                    self.use_type_name("qdev_create")
+                    if get_vp("use qdev_new"):
+                        self.use_type_name("qdev_new")
+                    else:
+                        self.use_type_name("qdev_create")
 
                     decl_code += "    DeviceState *%s;\n" % dev_name
 
-                    if ((node.parent_bus is None)
-                        or isinstance(node.parent_bus, SystemBusNode)
-                    ):
-                        bus_name = "NULL"
-                    else:
-                        bus_name = "BUS(%s)" % self.node_map[node.parent_bus]
-
                     if Type.exists(node.qom_type):
                         qom_type = node.qom_type
                     else:
                         qom_type = "\"%s\"" % node.qom_type
 
-                    def_code += """\
+                    if get_vp("use qdev_new"):
+                        def_code += """\
+    {dev_name} = qdev_new(@a{qom_type});{props_code}
+""".format(
+    dev_name = dev_name,
+    qom_type = qom_type,
+    props_code = props_code
+                        )
+                        if ((node.parent_bus is None)
+                            or isinstance(node.parent_bus, SystemBusNode)
+                        ):
+                            self.use_type_name("sysbus_realize_and_unref")
+                            self.use_type_name("SYS_BUS_DEVICE")
+                            def_code += """\
+    sysbus_realize_and_unref(@aSYS_BUS_DEVICE({dev_name}),@sNULL);
+""".format(
+    dev_name = dev_name
+                            )
+                        else:
+                            # TODO: test this branch using some non-sysbus
+                            #       configuration
+                            self.use_type_name("qdev_realize_and_unref")
+                            bus_name = \
+                                "BUS(%s)" % self.node_map[node.parent_bus]
+                            def_code += """\
+    qdev_realize_and_unref(@aSYS_BUS_DEVICE({dev_name}),@s{bus_name},@sNULL);
```

Comment:
I think that `qdev_realize_and_unref(@a{dev_name},@s{bus_name},@sNULL);`.

Reply:
dev_name is `DeviceState *` while `qdev_realize_and_unref` accepts `SysBusDevice *`. That's qemu, boy. =)

Reply:
`bool qdev_realize_and_unref(DeviceState *dev, BusState *bus, Error **errp)`
Why `SysBusDevice *`?

Reply:
Oops, I meant `sysbus_realize_and_unref`.